src/hal/src/external_memory/mod.rs

```diff
@@ -0,0 +1,498 @@
+//! Structures related to the import external memory functionality
+
+mod errors;
+pub use errors::*;
+
+pub use external_memory::*;
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub use drm_fourcc::*;
+
+/// Buffer or image
+#[derive(Debug)]
+pub enum Resource<B: crate::Backend> {
+    /// Buffer
+    Buffer(B::Buffer),
+    /// Image
+    Image(B::Image),
+}
+
+/// External memory properties
+#[derive(Debug, PartialEq)]
+pub struct ExternalMemoryProperties {
+    exportable: bool,
+    importable: bool,
+    exportable_from_imported: bool,
+}
+impl ExternalMemoryProperties {
+    /// Constructor
+    pub fn new(exportable: bool, importable: bool, exportable_from_imported: bool) -> Self {
+        Self {
+            exportable,
+            importable,
+            exportable_from_imported,
+        }
+    }
+    /// Is the queried configuration exportable
+    pub fn is_exportable(&self) -> bool {
+        self.exportable
+    }
+    /// Is the queried configuration importable
+    pub fn is_importable(&self) -> bool {
+        self.importable
+    }
+    /// Does the queried configuration requires dedicated allocation
+    pub fn is_exportable_from_imported(&self) -> bool {
+        self.exportable_from_imported
+    }
+}
+impl Default for ExternalMemoryProperties {
+    fn default() -> Self {
+        Self {
+            exportable: false,
+            importable: false,
+            exportable_from_imported: false,
+        }
+    }
+}
+
+bitflags!(
+    /// External memory type flags.
+    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+    pub struct ExternalMemoryTypeFlags: u32 {
```

Comment:
Can we move this into the `external-memory` crate? We'd want the gfx-hal interface to be platform-abstract, so the least we can do here is to encapsulate the `cfg()` stuff into a sub-crate.

Reply:
Yes, i can move it, or, if you prefer, i could just remove `cfg()`s and just expose every memory type to every platforms (and makes the functions panic if a memory type is used on the wrong platform). The initial idea was to just expose what the target platform can support, so that the user simply cannot use the wrong memory type, but i understant that having all these `cfg()`s could be undesirable.

Reply:
Moving seems fine